### PR TITLE
Make collection functions applicable to both arrays and objects

### PR DESCRIPTION
**Pull request by MonsieurV**
As mentioned in #35, we could make collection functions work on associative arrays but also objects.

This is a work in progress: tested on `__::pick()`, the next step is to add test cases with objects on each collection functions.
  
I let you know when it is to be reviewed for merge!

### CODE REVIEW

**Comment by allejo** [comment]
Don't mean to overstep here, but there seems to be a number of instances throughout the library checking whether it's an object or an array and it seems repetitive? Would it be helpful if there's a wrapper object introduced that's only used internally that implements `ArrayAccess`? That way, everything will be treated as an array instead of checking whether it's an object or an array?

**Comment by MonsieurV** [comment]
Yeah, that's a good point!

I won't make the argument to follow Lodash all the way. However that a good point may be to try to attain the consistency of its API and documentation.

The rationale of the PR is the following: (associative) arrays and objects are both first class citizens of PHP, so it makes sense to target the two in our collection functions.

For me it appears especially helpful to have functions as `map`, `reduce`, `has`, `set`, `pick` to work not only on arrays but also objects.

That is what I'm trying to do here. (hopefully not breaking too much the current API)

File: README.md [hunk 1]
```diff
@@ -94,6 +94,13 @@ __::compact([0, 1, false, 2, '', 3]);
 // >> [1, 2, 3]
 ```
 
+##### [__::drop](src/__/arrays/drop.php)
+Creates a slice of array with n elements dropped from the beginning.
+```php
+__::drop(1, 2, 3], 2);
```
Comment (allejo) [inline README.md]: missing a square brace for the array

Comment (MonsieurV) [reply README.md]: Thanks, fixed!

**Comment by MonsieurV** [comment]
I think I'm done here.

Not every collection function support objects, but we have the main ones.
There remains other still working uniquely on arrays, like `__::first` and `__::last` (we could argue that it makes no great sense to use them on objects anyway). Some would be great to convert when we can: `__::groupBy` for e.g.

It should fix #38

**Comment by MonsieurV** [comment]
@allejo Yeah, I agree that we should have limited occurrences on branching for object vs. array behavior.

I have the sentiment that when we have `__::get` and `__::set` implemented with both, we have the functional equivalent of the wrapper you're mentioning: we can return it in other function implementations.

However we still have other cases (that aren't just getting and setting property) when we need to do conditional branching: for e.g. to know which empty value to use following the collection. (https://github.com/YtoTech/bottomline/blob/9cf8d8ba394987f3000c67914623332644cc5e83/src/__/collections/set.php#L62)

To see if we have again and again these edge cases to manage, or if our `__::get` and `__:set` abstraction we're done about it

**Pull request merged**